### ISSUE TITLE
reduce level of halos around sources

### ISSUE THREAD

**Issue by twshimwell** [mhardcastle/ddf-pipeline#29]
To me smoothing the killms amplitudes looks very promising as it decrease the reduction in unmodelled flux (if e.g. no uvmin is used in the calibration) and it also substantially decreases the positive or negative halos around sources (see attached where left is without amplitude smoothing and right is with amplitude smoothing). 

Presently it seems to increase the noise a little but i need to investigate this is because I also see a noise increase if i set the smoothing to 1 (perhaps I'm messing up the phases a bit)

<img width="1338" alt="screen shot 2017-04-06 at 6 28 59 am" src="https://cloud.githubusercontent.com/assets/12065270/24737939/8756f028-1a92-11e7-8d98-da41cf1ea262.png">
<img width="1347" alt="screen shot 2017-04-06 at 6 28 45 am" src="https://cloud.githubusercontent.com/assets/12065270/24737941/87cbd82a-1a92-11e7-8654-c183b1c35da8.png">

**Comment by cyriltasse** [comment]
@twshimwell nice nice. Maybe you chould conduct the test without any normalisation to see the effects smoothing?

**Comment by twshimwell** [comment]
Ah ok yes good plan because this will also be impacted by the blbased normalisation. I'll try that and update again once done.

**Comment by twshimwell** [comment]
Turned off the normalisation and the noise looks a bit more comparable now (could probably do with a little more deconvolution but its not too bad). 

Here again the amplitude smoothing seems to have the benefit of reducing the positive/negative halos (whilst one facet in this image diverges here the amplitude smoothing also saves it a little bit). The right is with amplitude smoothing and the left is without amplitude smoothing. 

<img width="1225" alt="screen shot 2017-04-07 at 3 18 46 am" src="https://cloud.githubusercontent.com/assets/12065270/24781609/1e1267be-1b41-11e7-9466-f58eb738848b.png">
<img width="1223" alt="screen shot 2017-04-07 at 3 18 19 am" src="https://cloud.githubusercontent.com/assets/12065270/24781610/1e3a3f28-1b41-11e7-8016-3031c2c0ac6c.png">
<img width="1227" alt="screen shot 2017-04-07 at 3 25 04 am" src="https://cloud.githubusercontent.com/assets/12065270/24781727/e7517390-1b41-11e7-9b0f-a16e9f998f4d.png">

**Comment by twshimwell** [comment]
Any thoughts here?

The runs for comparison are:

/data/lofar/shimwell/L521348-v5-newkillms-ampsmooth 
/data/lofar/mjh/L521348-v5-newkillms/ (the image_full_ampphase1* needs redoing here as was run with the uvmax bug in the pipeline)

It is only the image_full_ampphase1* and the image_full_low* that have had the amplitude smoothing.

**Comment by mhardcastle** [comment]
How are you doing the amplitude smoothing? Are there scripts somewhere?

**Comment by twshimwell** [comment]
Its adapted from Cyrils Interpol.py script to just add a smoothing to the amplitudes.

/data/lofar/shimwell/L521348-v5-newkillms-ampsmooth/InterPol.py

This was just a fast way to do it so probably not the best. Cyril originally intended this script to read in all the killms files for all subbands and to smooth on them all together. However, perhaps we don't really need something like that and something simpler like this is ok.

Ideally though we don't need to do any of the interpolation in the InterPol.py script.

**Comment by mhardcastle** [comment]
Yeah, looks like we could simplify this massively to just smooth in time. Shall I have a look at doing that?

**Comment by twshimwell** [comment]
Great, yes that would be excellent if you have the time!

**Comment by mhardcastle** [comment]
In ddf-pipeline/master there is now a 'smoothsols.py' which just does the smoothing and not the interpolation -- it wasn't a big change in the end. The new code exposes the WSize parameter of the smooth which we probably need to experiment with. Have a look at this and if it seems sensible I can build the option to smooth with a given WSize into the pipeline. (Set --Plot=True to see the diagnostic plots.)

**Comment by twshimwell** [comment]
Great. I'll starting testing this tomorrow. I guess to test this we can:

-- test it on a few of the different normalisations for the full_low and full_high images -- so just smoothing the amps and running the final imaging
-- test a few different smoothings, perhaps 5min, 10 and 20min -- probably can again just do this on the final imaging.

**Comment by mhardcastle** [comment]
Yeah definitely -- test on a final image first. If it works well we should use it throughout the selfcal cycle though.

**Comment by mhardcastle** [comment]
I have a test implementation of this in the pipeline (branch smoothing -- for the full ampphase images only at the moment). Using it to test the effect on some of my fields.

**Comment by mhardcastle** [comment]
Just for the record -- optional smoothing of amp solutions throughout the selfcal is now in master.

**Comment by twshimwell** [comment]
Using the latest pipeline I have done the final imaging low resolution step without and then with a amplitude smoothing (smoothing = 53) on the pointing L562061. I see the noise on the low resolution image increase from 0.105mJy/beam to 0.11mJy/beam when smoothing the solutions. In this field the halos around sources decrease a bit (probably not completely dissapear) with the amplitude smoothing and the calibration quality doesnt look too different. The most clear example of the halos decreasing in this field is below with the smoothed on the left and unsmoothed on the right.

<img width="1358" alt="screen shot 2017-04-19 at 10 29 24 pm" src="https://cloud.githubusercontent.com/assets/12065270/25200906/5b6d28f6-2550-11e7-989a-eabeb72ce0b5.png">

I should get some of the high resolution images tomorrow and there are some fairly prominent halos in the unsmoothed image there.

**Comment by mhardcastle** [comment]
Will be worth checking the full-res image as that's the comparison I made. Also, if you form the difference image between the two above, what does it look like?

**Comment by twshimwell** [comment]
Whilst waiting for the full res ones. Here is a few difference images from the low res images (this is the smoothed minus unsmoothed) together with the smoothed  (top right) and unsmoothed (bottom left).


<img width="1053" alt="screen shot 2017-04-19 at 11 00 25 pm" src="https://cloud.githubusercontent.com/assets/12065270/25202041/312e42d8-2554-11e7-814e-0d1e85ac27d0.png">
<img width="1064" alt="screen shot 2017-04-19 at 11 00 08 pm" src="https://cloud.githubusercontent.com/assets/12065270/25202042/315481f0-2554-11e7-807d-0d766a8e945f.png">
<img width="749" alt="screen shot 2017-04-19 at 11 04 08 pm" src="https://cloud.githubusercontent.com/assets/12065270/25202129/8d4cc7e2-2554-11e7-8997-4623c8957224.png">

**Comment by mhardcastle** [comment]
Hm, not entirely sure what to make of that: but it's not simply affecting structure we don't believe in, is it? Is this done in combination with some sort of normalization? If so what sort?

**Comment by twshimwell** [comment]
Yes I'm not sure. The normalisation is the same with both being blbased. 

It may be because this is the difference between image_full_low.* so not the final deconvolved ones. The final ones should be ready by lunchtime so ill difference those too.

**Comment by twshimwell** [comment]
Any chance these blobs can simply be related to deconvolution rather than something more mysterious? In the attached I have all the image_full_low*app.restored.fits and image_full_low*model.fits (the model is shown both saturated and in a reasonable colour scale) and certainly for this extended source the diffuse blob of emission surrounding it is getting into the model (albeit at a very low level). 

The mask is also attached for all the image_full_low*mask* files (so all ones used at all stages in the cleaning I think).

Is there a parameter in the pipeline that controls how far clean is allowed to clean outside the masks?


<img width="820" alt="screen shot 2017-04-20 at 12 06 35 pm" src="https://cloud.githubusercontent.com/assets/12065270/25225516/f0d7cd76-25c1-11e7-8f79-1fc672eed525.png">
<img width="818" alt="screen shot 2017-04-20 at 12 05 41 pm" src="https://cloud.githubusercontent.com/assets/12065270/25225515/f0d3a5e8-25c1-11e7-897a-d704727f4786.png">
<img width="823" alt="screen shot 2017-04-20 at 12 07 22 pm" src="https://cloud.githubusercontent.com/assets/12065270/25225514/f0b5274e-25c1-11e7-80c8-ac10795fb830.png">

**Comment by mhardcastle** [comment]
In some cases I think extended haloes around extended sources are real -- or partly real. But doesn't explain the extended negative regions around some others.

I don't think we clean outside the masks at all.

**Comment by twshimwell** [comment]
I thought ddfacet can combine islands that are close to each other doesnt it? So in this case there are lots of the little islands and these combine to a bigger one.

**Comment by mhardcastle** [comment]
Does it? Not sure. @cyriltasse may have to enlighten us. Obviously automasking could do that in some circumstances.

In the low-res imaging we make our own masks for LSB stuff in between image_full_low_im and image_full_low_m . Those are convolved with a small kernel to avoid them having too many gaps in, and it's those that generally pick up your extended haloes.

**Comment by twshimwell** [comment]
I think I recall @cyriltasse mentioning something about a shape being placed around regions in the mask, see attached for example of mask, model and image. 

<img width="819" alt="screen shot 2017-04-20 at 12 29 29 pm" src="https://cloud.githubusercontent.com/assets/12065270/25226427/3720cf64-25c5-11e7-943f-85adebc24307.png">

No idea if this could be responsible though.

**Comment by twshimwell** [comment]
Hmm yes the low resolution fully cleaned image has finished now. The difference (this time unsmoothed minus smoothed) shows similar to previous.
<img width="819" alt="screen shot 2017-04-20 at 1 56 51 pm" src="https://cloud.githubusercontent.com/assets/12065270/25229664/653c9624-25d1-11e7-9366-a7f8e5f657e9.png">

The holes are about 0.5mJy

**Comment by twshimwell** [comment]
So, I think we are probably converging on this not being the solution :(

**Comment by mhardcastle** [comment]
Well, in that difference image it looks like signal is being taken out of haloes and put into the sources themselves, which might be what we want. Still, as you say, not the full solution perhaps, especially as it gives higher noise in full-res imaging in my tests.

I think (though I haven't yet tried on low-res images -- in progress) using SumBLBased rather than BLBased normalization is helping with haloes too.

**Comment by twshimwell** [comment]
so it is looking a bit encouraging this amplitude smoothing besides the noise increase. I wonder if worth following cyrils original plan and doing the smoothing of all solutions together (so over freq and time).

**Comment by mhardcastle** [comment]
Would be worth trying but for me the increased noise is a bit of a showstopper as things stand.

I think we need to define a set of steps and tests to address the blotchiness issue, but do that in parallel with some production runs to try to make progress on HETDEX.

**Comment by twshimwell** [comment]
Yes defining a set of steps and tests would be very useful. I'm not sure what the best approach is though.

Its interesting to see from the production runs how much this seems to vary better fields.  Looking at one of the fields running (P178+55) I am seeing very severe holes around sources (attached is the low_m image).

<img width="440" alt="screen shot 2017-04-25 at 5 10 32 pm" src="https://cloud.githubusercontent.com/assets/12065270/25392918/513363bc-29da-11e7-916f-1ff494611ec7.png">

**Comment by mhardcastle** [comment]
This issue seems to be superseded by the 'powersmooth' changes to killms so closing it for now.
